openhtf/io/http_api.py

```diff
@@ -34,16 +37,62 @@
 gflags.DEFINE_integer('http_port',
                       8888,
                       "Port on which to serve OpenHTF's HTTP API.")
+gflags.DEFINE_integer('multicast_port',
+                      10000,
+                      "Port on which to serve API port discovery service.")
+gflags.DEFINE_string('multicast_address',
+                     '239.1.1.1',
+                     'Address to use for API port discovery service.')
+
+
+_LOG = logging.getLogger(__name__)
+
+
+class AgentAlreadyListeningError(Exception):
+  """Multicast agents can't be used to send pings when listening."""
 
 
 class Server(util.threads.KillableThread):
-  """Bare-bones HTTP API server for OpenHTF.
+  """Frontend API server for openhtf.
 
+  Starts up two services. An HTTP server that serves detailed information about
+  this intance of openhtf (plus rundata for all instances on this host), and a
+  multicast server that helps frontends find and connect to the HTTP server.
+  
   Args:
     executor: An openhtf.exe.TestExecutor object.
   """
+  KILL_TIMEOUT_S = 1  # Seconds to wait between service kill attempts.
+
   def __init__(self, executor):
     super(Server, self).__init__()
+    self.servers = [HTTPServer(executor), MulticastAgent()]
+
+  def Start(self):
+    """Give the server a style-compliant Start method."""
```

Comment:
Yes, I would love to do this in KillableThread instead of all over the place. It's openhtf style (I think it's originally Google style -- as defined by our pylintrc) to differentiate methods (from other attributes in some cases, and from functions in other cases) by capitalizing them.